### PR TITLE
Properly specify Netcdf-hdf5 link interface

### PR DESCRIPTION
**Pull request by johnwparent**
Patch `netcdf-c` for proper hdf5 export; merge export patches

### CODE REVIEW

File: var/spack/repos/builtin/packages/netcdf-c/package.py [hunk 1]
```diff
@@ -342,6 +342,10 @@ def cmake_args(self):
         if "platform=windows" in self.pkg.spec:
             # Enforce the usage of the vendored version of bzip2 on Windows:
             base_cmake_args.append(self.define("Bz2_INCLUDE_DIRS", ""))
+        if "platform=windows" in self.pkg.spec and "+shared" in self.pkg.spec["hdf5"]:
+            base_cmake_args.append(self.define("NC_FIND_SHARED_LIBS", True))
+        else:
+            base_cmake_args.append(self.define("NC_FIND_SHARED_LIBS", False))
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/netcdf-c/package.py:351]: (question) To be clear, this was the default? `NC_FIND_SHARED_LIBS` should be off for `^hdf5+shared` on Linux etc.?

File: var/spack/repos/builtin/packages/netcdf-c/package.py [hunk 2]
```diff
@@ -59,8 +59,8 @@ class NetcdfC(CMakePackage, AutotoolsPackage):
         patch("4.8.1-win-hdf5-with-zlib.patch", when="@4.8.1: platform=windows")
 
         # TODO: fetch from the upstream repo once https://github.com/Unidata/netcdf-c/pull/2595
-        #  is accepted:
-        patch("netcdfc-mpi-win-support.patch", when="platform=windows")
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/netcdf-c/package.py:63]: It looks like nothing references this old patch now, so it should be removed.

**Review by scheibelp — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/netcdf-c/package.py (hunk 1 above)
Comment (johnwparent) [reply var/spack/repos/builtin/packages/netcdf-c/package.py:351]: It matters less for Linux, `hdf5` builds both static and shared by default, so regardless of the value, `netcdf-c` will find hdf5. This should probably respect the type requested by the user though, so I'll refactor to have the same check for both.

File: var/spack/repos/builtin/packages/netcdf-c/package.py (hunk 2 above)
Comment (johnwparent) [reply var/spack/repos/builtin/packages/netcdf-c/package.py:63]: Done.

**Comment by johnwparent** [comment]
@scheibelp requested changes addressed and pushed.

File: var/spack/repos/builtin/packages/netcdf-c/package.py (hunk 1 above)
Comment (scheibelp) [reply var/spack/repos/builtin/packages/netcdf-c/package.py:351]: Another question: does this mean that `netcdf-c` expects all dependencies to be built with shared libraries? If so, then perhaps we should in fact preserve whatever the default choice was on Linux (a comment about how hdf5 builds shared+static elsewhere would be useful in this case).

Comment (johnwparent) [reply var/spack/repos/builtin/packages/netcdf-c/package.py:351]: This variable is poorly named, it _only_ impacts `hdf5`'s `find_package` call, which is the reason I have the `hdf5` spec as part of the check. 
The default is whatever the default for `BUILD_SHARED_LIBS` is, which for Spack, is`netcdf-c`'s `shared` variant.

Comment (scheibelp) [reply var/spack/repos/builtin/packages/netcdf-c/package.py:351]: OK this sounds good as is - thanks!

File: var/spack/repos/builtin/packages/netcdf-c/package.py [hunk 3]
```diff
@@ -59,8 +59,8 @@ class NetcdfC(CMakePackage, AutotoolsPackage):
...
+        #  is accepted and the HDF5 interface patch is added
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/netcdf-c/package.py]: I'm reading this todo comment and getting confused:

* It sounds like https://github.com/Unidata/netcdf-c/pull/2595 is no longer sufficient
* "and the HDF5 interface patch is added" makes it sound like some other PR is required, but are the changes here not part of some PR to netcdf-c yet?

Taken together, I think the comment should be updated to something like:

```
# https://github.com/Unidata/netcdf-c/pull/2595 contains some of the changes in this patch but is not
# sufficient. There is no upstream PR that includes all these changes.
```

Comment (johnwparent) [reply var/spack/repos/builtin/packages/netcdf-c/package.py]: I upstreamed the MPI changes when I made them months ago, I just haven't had a chance to upstream the HDF5 related changes yet so there is no PR/set of PRs that encompasses the entire patch. 

Will update the comment.

**Review by scheibelp — APPROVED** [review]

**Pull request merged**